### PR TITLE
Use missing-prototypes in torch_cpu

### PR DESCRIPTION
**Pull request by cyyever**
This PR enables  Wmissing-prototypes in torch_cpu except some generated cpp files and the mps and metal backends.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @EikanWang

### CODE REVIEW

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -824,6 +824,21 @@ if(BUILD_CAFFE2 AND NOT MSVC)
   target_compile_options(torch_cpu PRIVATE "-Wno-sign-compare")
 endif()
 
+if("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
+  target_compile_options_if_supported(torch_cpu "-Wmissing-prototypes")
+  target_compile_options_if_supported(torch_cpu "-Werror=missing-prototypes")
+  set(ignored_files "${CMAKE_CURRENT_SOURCE_DIR}/utils/proto_wrap.cc")
+  foreach(generated_file IN LISTS GENERATED_CXX_TORCH ignored_files)
+    set_source_files_properties(${generated_file} PROPERTIES COMPILE_OPTIONS "-Wno-missing-prototypes;-Wno-error=missing-prototypes")
```
Comment (albanD) [inline caffe2/CMakeLists.txt:832]: Is the plan to fix these in a follow up? Or is it the expected end state?

**Review by albanD — APPROVED** [review]
very nice cleanup!

File: caffe2/CMakeLists.txt (hunk 1 above)
Comment (cyyever) [reply caffe2/CMakeLists.txt:832]: No plan to fix them now.

**Comment by huydhn** [comment]
@cyyever It looks like this PR hasn't covered iOS job in periodic, so it's failing atm with `error: no previous prototype for function 'pack_outputs' [-Werror,-Wmissing-prototypes]` https://hud.pytorch.org/pytorch/pytorch/commit/716b3b893d2826f1e47ab5321f082b48c66c8c92

Could you help take a look and let me know if this should be forward fixed for reverted?

**Comment by osalpekar** [comment]
This is a great cleanup, but looks like making two of the operators in caffe2 static caused linker errors when building caffe2:
```
[2023-06-21T17:45:52.127-07:00] Stderr: ld.lld: error: undefined symbol: caffe2::TensorInferenceForBatchMatMul(caffe2::OperatorDef const&, std::vector<caffe2::TensorShape, std::allocator<caffe2::TensorShape> > const&)
[2023-06-21T17:45:52.127-07:00] ld.lld: error: undefined symbol: caffe2::CostInferenceForBatchMatMul(caffe2::OperatorDef const&, std::vector<caffe2::TensorShape, std::allocator<caffe2::TensorShape> > const&)
```

Unfortunately we don't build caffe2 in the CI run per-PR. Perhaps it makes sense for somebody to shepherd this change through? cc @albanD

**Comment by cyyever** [comment]
@osalpekar I can't reproduce locally, can you provide the build commands?

**Comment by cyyever** [comment]
@huydhn  I would fix it

**Pull request closed without merging**